File: api_py/service/sqlite_record_service.py
import json

import aiosqlite

from entity.record import Record
from service.record_service import RecordService, ServiceError
from util.log import log_error
from service.queries import *

class SQLiteRecordService(RecordService):
    """SQLite-backed implementation of RecordService.

    Example:
        service = SQLiteRecordService("timetravel.db")
    """

    def __init__(self, path: str) -> None:
        self._path = path
        self._db: aiosqlite.Connection | None = None

    async def _get_db(self) -> aiosqlite.Connection:
        """Returns the open DB connection, initialising it on first call."""
        if self._db is None:
            try:
                self._db = await aiosqlite.connect(self._path)
                await self._db.execute(queries.get('v1').get(CREATE_TABLE_RECORDS))
                await self._db.execute(queries.get('v2').get(CREATE_TABLE_RECORDS))
                await self._db.commit()
            except Exception as err:
                log_error(err)
                raise
        return self._db

    async def get_record(self, id: int) -> Record | None:
        try:
            db = await self._get_db()
            async with db.execute(queries.get('v1').get(GET_RECORD_BY_ID), (id,)) as cursor:
                row = await cursor.fetchone()
        except Exception as err:
            log_error(err)
            raise

        if row is None:
            return None

        try:
            data: dict[str, str] = json.loads(row[0])
        except Exception as err:
            log_error(err)
            raise

        return Record(id=id, data=data)

    async def create_record(self, record: Record) -> None:
        if record.id <= 0:
            err = ServiceError.invalid_id()
            log_error(err)
            raise err

        try:
            data_json = json.dumps(record.data)
            db = await self._get_db()
            cursor = await db.execute(
                queries.get('v1').get(INSERT_RECORD),
                (record.id, data_json),
            )
            await db.commit()
            rows_affected = cursor.rowcount
        except Exception as err:
            log_error(err)
            raise

        if rows_affected == 0:
            err = ServiceError.already_exists()
            log_error(err)
            raise err
        return rows_affected

    async def update_record(self, id: int, updates: dict[str, str | None]) -> Record:
        record = await self.get_record(id)
        if record is None:
            raise ServiceError.not_found()

        #TODO: Is it better to have a fallback to create if record doesn't exist
        for key, value in updates.items():
            if value is None:  # deletion update
                record.data.pop(key, None)
            else:
                record.data[key] = value

        try:
            data_json = json.dumps(record.data)
            db = await self._get_db()
            await db.execute(
                queries.get('v1').get(UPDATE_RECORD),
                (data_json, id),
            )
            await db.commit()
            db.close()
        except Exception as err:
            log_error(err)
            raise

        return record

    async def get_record_history(self) -> list[Record]:
        try:
            with await self._get_db() as conn:
                cursor = conn.execute(queries.get('v1').get(GET_RECORD_HISTORY), (id))
                rows = cursor.fetchall()
                return [Record.model_validate_json(record) for record in rows]
        except Exception as err:
            log_error(err)
            raise
        return

    async def get_latest_record_version(self):
        pass

    async def get_record_version(self):
        pass

File: service/sqlite_record_service.go
package service

import (
	"context"
	"database/sql"
	"encoding/json"
	"timetravel/entity"
	"timetravel/util"

	_ "modernc.org/sqlite"
)

// SQLiteRecordService is a SQLite-backed implementation of RecordService.
type SQLiteRecordService struct {
	db *sql.DB
}

func NewSQLiteRecordService(path string) SQLiteRecordService {
	db, err := sql.Open("sqlite", path)
	if err != nil {
		util.LogError(err)
		return SQLiteRecordService{}
	}

	_, err = db.Exec(`CREATE TABLE IF NOT EXISTS records (id INTEGER PRIMARY KEY, data TEXT NOT NULL)`)
	if err != nil {
		util.LogError(err)
		return SQLiteRecordService{}
	}

	return SQLiteRecordService{db: db}
}

func (s *SQLiteRecordService) GetRecord(ctx context.Context, id int) (entity.Record, error) {
	row := s.db.QueryRowContext(ctx, `SELECT data FROM records WHERE id = ?`, id)

	var dataJSON string
	if err := row.Scan(&dataJSON); err != nil {
		if err == sql.ErrNoRows {
			util.LogError(ErrRecordDoesNotExist)
			return entity.Record{}, ErrRecordDoesNotExist
		}
		util.LogError(err)
		return entity.Record{}, err
	}

	var data map[string]string
	if err := json.Unmarshal([]byte(dataJSON), &data); err != nil {
		util.LogError(err)
		return entity.Record{}, err
	}

	return entity.Record{ID: id, Data: data}, nil
}

func (s *SQLiteRecordService) CreateRecord(ctx context.Context, record entity.Record) error {
	id := record.ID
	if id <= 0 {
		util.LogError(ErrRecordIDInvalid)
		return ErrRecordIDInvalid
	}

	dataJSON, err := json.Marshal(record.Data)
	if err != nil {
		util.LogError(err)
		return err
	}

	result, err := s.db.ExecContext(ctx, `INSERT OR IGNORE INTO records (id, data) VALUES (?, ?)`, id, string(dataJSON))
	if err != nil {
		util.LogError(err)
		return err
	}

	rows, err := result.RowsAffected()
	if err != nil {
		util.LogError(err)
		return err
	}
	if rows == 0 {
		util.LogError(ErrRecordAlreadyExists)
		return ErrRecordAlreadyExists
	}

	return nil
}

func (s *SQLiteRecordService) UpdateRecord(ctx context.Context, id int, updates map[string]*string) (entity.Record, error) {
	record, err := s.GetRecord(ctx, id)
	if err != nil {
		return entity.Record{}, err
	}

	for key, value := range updates {
		if value == nil { // deletion update
			delete(record.Data, key)
		} else {
			record.Data[key] = *value
		}
	}

	dataJSON, err := json.Marshal(record.Data)
	if err != nil {
		util.LogError(err)
		return entity.Record{}, err
	}

	_, err = s.db.ExecContext(ctx, `UPDATE records SET data = ? WHERE id = ?`, string(dataJSON), id)
	if err != nil {
		util.LogError(err)
		return entity.Record{}, err
	}

	return record, nil
}
